x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotAllocator.java

```diff
@@ -103,16 +215,239 @@ public AllocateUnassignedDecision explainUnassignedShardAllocation(ShardRouting
+            DiscoveryNode discoNode = nodeStoreEntry.getKey();
+            NodeCacheFilesMetadata nodeCacheFilesMetadata = nodeStoreEntry.getValue();
+            // we don't have any existing cached bytes at all
+            if (nodeCacheFilesMetadata.bytesCached() == 0L) {
+                continue;
+            }
+
+            RoutingNode node = allocation.routingNodes().node(discoNode.getId());
+            if (node == null) {
+                continue;
+            }
+
+            // check if we can allocate on the node
+            Decision decision = allocation.deciders().canAllocate(shard, node, allocation);
+            Long matchingBytes = null;
+            if (explain) {
+                matchingBytes = nodeCacheFilesMetadata.bytesCached();
+                NodeAllocationResult.ShardStoreInfo shardStoreInfo = new NodeAllocationResult.ShardStoreInfo(matchingBytes);
+                nodeDecisionsDebug.put(node.nodeId(), new NodeAllocationResult(discoNode, shardStoreInfo, decision));
+            }
+
+            if (decision.type() == Decision.Type.NO) {
+                continue;
+            }
+
+            if (matchingBytes == null) {
+                matchingBytes = nodeCacheFilesMetadata.bytesCached();
+            }
+            matchingNodesCacheSizes.put(discoNode, matchingBytes);
+            if (logger.isTraceEnabled()) {
+                logger.trace(
+                    "{}: node [{}] has [{}/{}] bytes of re-usable cache data",
+                    shard,
+                    discoNode.getName(),
+                    new ByteSizeValue(matchingBytes),
+                    matchingBytes
+                );
+            }
+        }
+
+        return new MatchingNodes(matchingNodesCacheSizes, nodeDecisionsDebug);
+    }
+
+    private static final class AsyncCacheStatusFetch {
+
+        private final Set<DiscoveryNode> fetchingDataNodes = new HashSet<>();
+
+        private final Map<DiscoveryNode, NodeCacheFilesMetadata> data = new HashMap<>();
+
+        AsyncCacheStatusFetch() {}
+
+        synchronized DiscoveryNode[] addFetches(DiscoveryNode[] nodes) {
+            final Collection<DiscoveryNode> nodesToFetch = new ArrayList<>();
+            for (DiscoveryNode node : nodes) {
+                if (data.containsKey(node) == false && fetchingDataNodes.add(node)) {
+                    nodesToFetch.add(node);
+                }
+            }
+            return nodesToFetch.toArray(new DiscoveryNode[0]);
+        }
+
+        synchronized void addData(Map<DiscoveryNode, NodeCacheFilesMetadata> newData) {
+            data.putAll(newData);
+            fetchingDataNodes.removeAll(newData.keySet());
+        }
+
+        @Nullable
+        synchronized Map<DiscoveryNode, NodeCacheFilesMetadata> data() {
+            return fetchingDataNodes.size() > 0 ? null : Map.copyOf(data);
+        }
+
+        synchronized int numberOfInFlightFetches() {
```

Comment:
There might be better (as in more efficient) ways (certainly are) to build the synchronization here but I decided to keep it as simple as possible in the interest of saving some time today.